src/Workspaces/Core/Portable/Workspace/Workspace_Editor.cs

```diff
@@ -455,16 +441,44 @@ private ISet<DocumentId> GetProjectOpenDocuments_NoLock(ProjectId project)
                 var newDoc = newSolution.GetDocument(documentId);
                 this.OnDocumentTextChanged(newDoc);
 
-                // Fire and forget that the workspace is changing
-                var ignore1 = RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
-                var ignore2 = this.RaiseDocumentOpenedEventAsync(newDoc); // don't await this
+                // Fire and forget that the workspace is changing.
+                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
+                this.RaiseDocumentOpenedEventAsync(newDoc);
             }
-            finally
+
+            this.RegisterText(textContainer);
+        }
+
+        private class ReuseVersionLoader : TextLoader
+        {
+            // Capture DocumentState instead of Document so that we don't hold onto the old solution.
+            private readonly DocumentState _oldDocumentState;
+            private readonly SourceText _newText;
+
+            public ReuseVersionLoader(DocumentState oldDocumentState, SourceText newText)
             {
-                disposable.Dispose();
+                _oldDocumentState = oldDocumentState;
```

Comment:
I think that one of the concerns that we had with doing something like this was ending up with long chains of loaders like this that held every snapshot alive.